### PR TITLE
gh-78502: Add a trackfd parameter to mmap.mmap()

### PR DESCRIPTION
**Pull request by ZackerySpytz**
If *trackfd* is False, the file descriptor specified by *fileno*
will not be duplicated.

https://bugs.python.org/issue34321

* Issue: gh-78502

### CODE REVIEW

**Comment by MojoVampire** [comment]
Could you make the new trackfd argument keyword-only? When the number of arguments starts getting large, especially when it's an argument that's boolean and not self-describing, it seems like there is a preference for making the new arguments keyword-only both to prevent using the API in needlessly unclear ways and to avoid a long term compatibility requirement that prevents changing the argument order (see evolution of subprocess.Popen and subprocess.run). I suspect, in hindsight, mmap would have liked to keep more arguments keyword-only to avoid the issue with differing argument orders on UNIX-likes and Windows, but that ship has sailed; at least we can avoid making it any worse.

Also, is there a reason you're not doing this for Windows? The same feature seems perfectly reasonable in Windows too.

**Comment by erlend-aasland** [comment]
cc. @encukou / @serhiy-storchaka: would one of you like to take a look at the proposed solution?

**Comment by erlend-aasland** [comment]
> This is a good start! @ZackerySpytz, do you want to continue or should I finish the feature?

Zackery is not very active these days; I think it is fine to take over the PR.

**Comment by encukou** [comment]
OK! Here are my updates.

I've realized that `resize()` method works fine without a `fd` -- it'll resize the map, but not the underlying file, so if the map grows, the “new” part becomes unmapped.
The latest commit here fixes that -- at the cost of an additional flag. (The *memory* cost in negligible IMO -- and could be optimized away -- but there's a maintenance cost & mental overhead.)
I'm not sure if `resize()` with `trackfd=False` would ever be useful. Making it fail is a safer initial implementation; it can always be enabled.

File: Doc/library/mmap.rst
Comment (serhiy-storchaka) [inline Doc/library/mmap.rst]: `~` and `!` do not work together.
```suggestion
   not be duplicated, and the resulting :class:`!mmap` object will not
```

File: Doc/library/mmap.rst
Comment (serhiy-storchaka) [inline Doc/library/mmap.rst]: Maybe make a new optional boolean parameter keyword-only?

File: Lib/test/test_mmap.py
Comment (serhiy-storchaka) [inline Lib/test/test_mmap.py]: What is errno?

File: Modules/mmapmodule.c
Comment (serhiy-storchaka) [inline Modules/mmapmodule.c]: Shouldn't it be ValueError?

**Review by serhiy-storchaka — COMMENTED** [review]
I think that it would be interesting to add the following tests:

* Test that it fails on Windows with `trackfd=False` and `trackfd=True`.
* Test with `fd=-1`.
* Test with original `fd` closed after creating a mmap.
* Test whether mmap can be used after fork.

File: Modules/mmapmodule.c
Comment (encukou) [reply Modules/mmapmodule.c]: It's `TypeError` for consistency with the `(self->access == ACCESS_WRITE) || (self->access == ACCESS_DEFAULT)` case.

File: Lib/test/test_mmap.py
Comment (encukou) [reply Lib/test/test_mmap.py]: EBADF, directly from the system. I can add it to the test.

File: Doc/library/mmap.rst
Comment (encukou) [reply Doc/library/mmap.rst]: They work fine for Python.
Sphinx has (had?) a bug where they don't work together for the C domain.

File: Doc/whatsnew/3.10.rst
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.10.rst]:
```suggestion
(Contributed by Zackery Spytz and Petr Viktorin  in :issue:`34321`.)
```

And move this to 3.13.

File: Lib/test/test_mmap.py
Comment (serhiy-storchaka) [inline Lib/test/test_mmap.py]: Does it work the same for `mmap(-1, size, trackfd=True)`? I have not found tests for this.

File: Modules/mmapmodule.c
Comment (serhiy-storchaka) [reply Modules/mmapmodule.c]: Usually ValueError is raised if the object is in the wrong state (for example when write to a closed file).

Comment (encukou) [reply Modules/mmapmodule.c]: That makes the docs cumbersome, but it makes sense.

File: Doc/whatsnew/3.10.rst
Comment (encukou) [reply Doc/whatsnew/3.10.rst]: Thanks for the catch!

File: Lib/test/test_mmap.py
Comment (encukou) [reply Lib/test/test_mmap.py]: `trackfd=True` is the default. It's tested in `test_resize_past_pos`.

File: Doc/library/mmap.rst
Comment (serhiy-storchaka) [reply Doc/library/mmap.rst]: It is not the C domain, but Sphinx emits a warning.

**Review by serhiy-storchaka — COMMENTED** [review]
By default, the mmap is always created from non-inheritable file descriptor. But if trackfd is false, it can be created from inheritable file descriptor. I wondered what is the difference and how can it be tested.

But if it works after closing the original file descriptor, perhaps there is no difference. cc @vstinner

File: Lib/test/test_mmap.py
Comment (serhiy-storchaka) [reply Lib/test/test_mmap.py]: It has different behavior. Should `trackfd=False` have any effect for `fd=-1`? Should combination `fd=-1` and `trackfd=False` be allowed?

Comment (encukou) [reply Lib/test/test_mmap.py]: Conceptually, `trackfd=False` should be the only possible behaviour for `fd=-1` -- the fd is not tracked.
However, I don't think it's worth it to include a third value for “trackfd=*default*”. As it is, the default (`True`) works as before. Setting `False` has the documented effect -- disabling `resize`. This isn't very useful, but I don't think it's worth handling specially.

**Comment by encukou** [comment]
There's always another test that could be added. I think we're too far into diminishing returns at this point :)

**Comment by serhiy-storchaka** [comment]
There is currently a discussion about inheritance of file descriptors in #113817. I just want to be sure that we do not miss some details here (most likely it is all correct). cc @vstinner, @zooba, @eryksun.

File: Doc/library/mmap.rst
Comment (zooba) [inline Doc/library/mmap.rst:106]: I'd love to have some idea of why I might want to use this parameter. Right now it only describes the downsides.

**Review by zooba — COMMENTED** [review]
Our other discussion was specific to Windows, so this issue doesn't apply.

But since I got mentioned, could we get the docs phrased in a way that explains why I might ever want to use the new option? Right now it sounds like a bad thing.

File: Doc/library/mmap.rst
Comment (eryksun) [reply Doc/library/mmap.rst:106]: On Windows, the internally duplicated handle probably references an open that lacks delete access. It thus prevents deleting the file, even if the mapped section otherwise allows it (e.g. the section is mapped readonly). For example:

```pycon
>>> f = open('spam.txt')
>>> m = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
>>> f.close()
>>> os.remove('spam.txt')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'spam.txt'

>>> # I manually closed the internal handle via Process Explorer.
>>> os.remove('spam.txt')
>>> m[:]
b'spam'
```

Comment (zooba) [reply Doc/library/mmap.rst:106]: That sounds like a reason to at least add the argument for all platforms, which I'm generally in favour of anyway. It can have more appropriate semantics on Windows if needed (i.e. "doesn't hold an extra HANDLE" rather than "FD").

It's probably actually pretty useful to be able to immediately delete the file but keep the mapping open (which will keep the file on disk on Windows at least, so you can't reuse the name while it's in use). And it looks like the mapping doesn't lock out deletes, so I guess it'll work as intended.

I'm not going to hold up this PR for it though. All I'll say is that if we ever do add that option, it should be `trackfd=False` to "activate" it, for consistency between platforms.

Comment (eryksun) [reply Doc/library/mmap.rst:106]: > That sounds like a reason to at least add the argument for all platforms, which I'm generally in favour of anyway. It can have more appropriate semantics on Windows if needed (i.e. "doesn't hold an extra HANDLE" rather than "FD").

I think `trackfd` would be fine on Windows. The `fileno` parameter is a C file descriptor, not a native OS handle.

> It's probably actually pretty useful to be able to immediately delete the file but keep the mapping open (which will keep the file on disk on Windows at least, so you can't reuse the name while it's in use). And it looks like the mapping doesn't lock out deletes, so I guess it'll work as intended.

NTFS supports POSIX delete, in which a deleted file gets renamed to a reserved system directory until all references to the file object have been closed. That includes the internal pointer reference to a file object that's held by the memory manager for the mapped section. The internal file reference doesn't count toward the file's share mode, i.e. a memory-mapped file can be deleted even if the source open didn't share delete access. Actually, I just checked that the delete is allowed nowadays even if the mapped section has write access to the file, so my assumption was wrong that it would only work for a readonly mapping.

You can observe this in Process Explorer. Switch the lower-pane view to DLLs (file- and pagefile-backed memory mappings), and add the name and path columns to the view. You'll see that the backing file gets moved to the "\\$Extend\\$Deleted" system directory on the volume after the file is 'deleted'.

**Comment by encukou** [comment]
I'd prefer if the Windows functionality was added in a separate PR. I don't have a Windows box set up (yet), so I won't send one myself.
Other than that, if the added note is OK this looks good to merge.

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM.

**Review by serhiy-storchaka — APPROVED** [review]

File: Doc/library/mmap.rst
Comment (chris6090) [reply Doc/library/mmap.rst]: Kriseng83@gmail.com

**Pull request merged**
